### PR TITLE
TRUNK-6185: Easier configuration of csrfguard.properties

### PR DESCRIPTION
**Pull request by wikumChamith**
## Description of what I changed

This commit updates the loadCsrfGuardProperties method to enhance the loading of CSRF Guard properties. It first checks for the existence of the csrfguard.properties file in the `.OpenMRS` directory. If the file exists, its contents are loaded. Otherwise, the default values are loaded from the `/WEB-INF/csrfguard.properties` file. Additionally, properties starting with `org.owasp.csrfguard` from `openmrs-runtime.properties` are merged into the loaded properties. 

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6185

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Review by mseaton — CHANGES_REQUESTED** [review]

**Comment by wikumChamith** [comment]
@mseaton I addressed the issues you pointed out.

**Comment by mseaton** [comment]
@wikumChamith , no this doesn't look right.  You shouldn't be putting logic into the catch block, and you still don't have any finally blocks where you close the input streams.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 1]
```diff
@@ -258,29 +259,34 @@ public void contextInitialized(ServletContextEvent event) {
...
-		    
-		    //Moved to EOF by the copy operation. So open it again.
-		    inputStream = new FileInputStream(file);
-		}
-		Properties properties = new Properties();
-		properties.load(inputStream);
-		IOUtils.closeQuietly(inputStream);
-		CsrfGuard.load(properties);
-		
+			try {
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/Listener.java:272]: Doesn't seem right to do this _unconditionally_ in the catch block. There are numerous reasons an `IOException` could be thrown above, but only one where we'd want to fallback: a `FileNotFound` exception. On top of which, the error here is always swallowed.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 2]
```diff
@@ -258,29 +259,34 @@ public void contextInitialized(ServletContextEvent event) {
...
+				csrfGuardInputStream = Files.newInputStream(Paths.get(fileName));
+			} 
+			catch (IOException ioException1) {
+				ioException1.printStackTrace();
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/Listener.java:276]: Log, rather than printing.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 3]
```diff
@@ -258,29 +259,34 @@ public void contextInitialized(ServletContextEvent event) {
...
+				throw ioException1;
+			}
+		}
+		csrfGuardProperties.load(csrfGuardInputStream);
+		Properties runtimeProperties = OpenmrsUtil.getRuntimeProperties(WebConstants.WEBAPP_NAME);
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/Listener.java:281]: It's preferable to use `Context.getRuntimeProperties()`

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 4]
```diff
@@ -258,29 +259,34 @@ public void contextInitialized(ServletContextEvent event) {
...
+		runtimeProperties.stringPropertyNames().forEach(property -> {
+			if(property.startsWith("org.owasp.csrfguard")) {
+				csrfGuardProperties.setProperty(property, runtimeProperties.getProperty(property));
+			}
+		});		
+		IOUtils.closeQuietly(csrfGuardInputStream);
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/Listener.java:287]: This really needs to be done in a finally block. There are way to many ways an exception can be thrown in the above code and so the IO stream never closed.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 5]
```diff
@@ -258,29 +259,34 @@ public void contextInitialized(ServletContextEvent event) {
 			log.error(MarkerFactory.getMarker("FATAL"), "Failed to obtain JDBC connection", e);
 		}
 	}
-	
-	private void loadCsrfGuardProperties(ServletContext servletContext) throws FileNotFoundException, IOException {	
-		File file = new File(OpenmrsUtil.getApplicationDataDirectory(), "csrfguard.properties");
-		InputStream inputStream = null;
+
+	private void loadCsrfGuardProperties(ServletContext servletContext) throws IOException {
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/Listener.java:263]: Note that `loadCsrfGuardProperties()` is currently called _before_ the runtime properties are setup, which means `OpenmrsUtil.getApplicationDirectory()` (inter alia) can return the wrong folder. Ideally, this call should be moved to after the runtime properties are loaded.

**Comment by wikumChamith** [comment]
@mseaton, @ibacher I refactored the code according to your feedback.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 6]
```diff
@@ -258,28 +257,35 @@ public void contextInitialized(ServletContextEvent event) {
...
+		File csrfGuardFile = new File(OpenmrsUtil.getApplicationDataDirectory(), "csrfguard.properties");
+		Properties csrfGuardProperties = new Properties();
+		if(csrfGuardFile.exists()) {
+			try(InputStream csrfGuardInputStream = Files.newInputStream(csrfGuardFile.toPath())) {
+				csrfGuardProperties.load(csrfGuardInputStream);
+			}
+			catch (Exception e) {
+				log.error(e.getMessage());
```
Comment (mseaton) [inline web/src/main/java/org/openmrs/web/Listener.java:269]: I'm not sure about just logging an error here and continuing.  Why catch this Exception if you aren't going to fall back to other behavior?  If we want to log this here (which is probably a good idea), then I'd add some additional context in the logging message to make it easier for someone to know where the problem is and what went wrong, and then re-throw the Exception.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 7]
```diff
@@ -258,28 +257,35 @@ public void contextInitialized(ServletContextEvent event) {
...
+			}
+		}
+		else {
+			String fileName = servletContext.getRealPath("/WEB-INF/csrfguard.properties");
+			try (InputStream csrfGuardInputStream = Files.newInputStream(Paths.get(fileName))) {
+				csrfGuardProperties.load(csrfGuardInputStream);
+			}
+			catch (Exception e) {
+				log.error(e.getMessage());
```
Comment (mseaton) [inline web/src/main/java/org/openmrs/web/Listener.java:278]: Same comment here.  Just log, with a more helpful message for implementers to better understand the error, and then re-throw the Exception

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 8]
```diff
@@ -258,28 +257,35 @@ public void contextInitialized(ServletContextEvent event) {
...
+			}
+		}
+		
+		Properties runtimeProperties = getRuntimeProperties();
+		runtimeProperties.stringPropertyNames().forEach(property -> {
+			if(property.startsWith("org.owasp.csrfguard")) {
```
Comment (mseaton) [inline web/src/main/java/org/openmrs/web/Listener.java:284]: Style comment - please put a space between the `if` and the `(`.  Apply this elsewhere in this PR as well.

**Review by mseaton — CHANGES_REQUESTED** [review]
Looking much better - a few additional comments to consider.

**Comment by wikumChamith** [comment]
@mseaton I addressed the issues you pointed out. Please review the new code.

**Review by mseaton — COMMENTED** [review]
Just a few minor style comments @wikumChamith and then I think it looks good.  The only other thing that I'd love to see are some unit tests around this, if this is possible to do.  Not sure if @ibacher has any additional feedback.

**Comment by wikumChamith** [comment]
@mseaton I addressed the suggestions you made. Any suggestions on a viable way that I can write tests for this method?

**Review by mseaton — APPROVED** [review]

**Pull request merged**